idom_router/core.py

```diff
@@ -0,0 +1,142 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, replace
+from pathlib import Path
+from typing import Any, Callable, Iterator, Sequence, TypeVar
+from urllib.parse import parse_qs
+
+from idom import (
+    component,
+    create_context,
+    use_context,
+    use_location,
+    use_memo,
+    use_state,
+)
+from idom.backend.hooks import ConnectionContext, use_connection
+from idom.backend.types import Connection, Location
+from idom.core.types import VdomChild, VdomDict
+from idom.types import ComponentType, Context, Location
+from idom.web.module import export, module_from_file
+
+from idom_router.types import Route, RouteCompiler, RouteResolver, Router
+
+R = TypeVar("R", bound=Route)
+
+
+def create_router(compiler: RouteCompiler[R]) -> Router[R]:
+    """A decorator that turns a route compiler into a router"""
+
+    def wrapper(*routes: R) -> ComponentType:
+        return router_component(*routes, compiler=compiler)
+
+    return wrapper
+
+
+@component
+def router_component(
+    *routes: R,
+    compiler: RouteCompiler[R],
+) -> ComponentType | None:
+    old_conn = use_connection()
+    location, set_location = use_state(old_conn.location)
+    router_state = use_context(_route_state_context)
+
+
+    if router_state is not None:
+        raise RuntimeError("Another router is already active in this context")
+
+    # Memoize the compiled routes and the match separately so that we don't
+    # recompile the routes on renders where only the location has changed
+    compiled_routes = use_memo(lambda: _compile_routes(routes, compiler))
+    match = use_memo(lambda: _match_route(compiled_routes, location))
+
+    if match is not None:
+        route, params = match
+        return ConnectionContext(
+            _route_state_context(
+                route.element, value=_RouteState(set_location, params)
+            ),
+            value=Connection(old_conn.scope, location, old_conn.carrier),
+            key=route.path,
+        )
+
+    return None
+
+
+@component
+def link(*children: VdomChild, to: str, **attributes: Any) -> VdomDict:
+    set_location = _use_route_state().set_location
+    attrs = {
+        **attributes,
+        "to": to,
+        "onClick": lambda event: set_location(Location(**event)),
+    }
+    return _link(attrs, *children)
+
+
+def use_params() -> dict[str, Any]:
+    """Get parameters from the currently matching route pattern"""
+    return _use_route_state().params
+
+
+def use_query(
```

Comment:
Could be confusing with Django's `use_query` hook? I'm conflicted on this.

Reply:
I think `reactpy-django` should just re-export things to avoid confusion.

Reply:
Can rename this to `use_search`

Reply:
Maybe `use_query_string` is also different enough?

Reply:
This was technically present prior to these changes so I think we can rename in a follow-up.